sqlitecluster/SQLiteNode.cpp

```diff
@@ -74,6 +80,143 @@ SQLiteNode::~SQLiteNode() {
     SASSERTWARN(!commitInProgress());
 }
 
+void SQLiteNode::replicate(SQLiteNode& node, SQLite& db, int threadNum) {
+    SInitialize("repl" + to_string(threadNum));
+    auto _state = node._state.load();
+    string name = "xxx";
+
+    // If we're already in the middle of handling a transaction, we'll set this to true. This allows us to re-use the
+    // main loop and wait logic.
+    // 0 = no command.
```

Comment:
I feel like there's got to be a way to structure this to not require the ENUM.  Basically, each worker thread fully processes one message, so I don't think it needs this idea of transaction state.  What about:

```
while (!node._replicationThreadsShouldExit) {
    command = node._replicationCommands.pop();
    if (command == "BEGIN_TRANSACTION") {
        // Wait for it to be our turn
        while (db.commitCount < command.commitCount-1) {
            _replicationCV.wait();
            if (node._replicationThreadsShouldExit) {
                // Can't begin this transaction because we're longer following
                return;
            }
        }

        // It's our turn!  Begin the transaction:
        db.begin(command.sql);

        // Now wait to see if we can commit or not
        while (true) {
            // Can we commit?
            if (hash in _replicationHashesToRollback) {
                db.rollback();
                break;
            } else if (hash in _ replicationHashesToCommit) {
                db.commit();
                break;
            }

            // Keep waiting
            _replicationCV.wait();
            if (node._replicationThreadsShouldExit) {
                // Can't commit or rollback this transaction because we're longer following
                return;
            }
        }
    } else if (command == "ROLLBACK_TRANSACTION") {
        _replicationHashesToRollback.insert(command.hash);
    } else if (command == "COMMIT_TRANSACTION") {
        _replicationHashesToCommit.insert(command.hash);
    } else {
        SERROR("Unknown replication command: " + command);
    }
}
```

Reply:
Actually... does this have a starvation issue where all 8 replication threads might be blocked waiting to commit a transaction they've begun, such that nobody is available to receive the COMMIT/ROLLBACK_COMMAND message?  Given that the ROLLBACK/COMMIT_TRANSACTION logic is so minimal, maybe those don't need to be multithreaded at all -- just have the sync thread handle the logic of updating the _replicationHashesToCommit/Rollback sets, and then the replication threads *only* handle BEGIN_TRANSACTION commands (which is the only thing we care about multithreading).  This way even if we get 1000 BEGIN_TRANSACTIONS in a row before getting a single COMMIT_TRANSACTION, it'll never get clogged and deadlocked.

Reply:
There's no starvation condition currently so long as there are at least two threads. This is only true because messages are always enqueued in order - leader never sends a `COMMIT` before it sends a `BEGIN` for the same transaction. This will not necessarily be true in a future parallel networking version of bedrock, but is true now.

Example:
Thread 1 does `BEGIN`, and blocks waiting on `COMMIT` (or `ROLLBACK`).

Thread 2 dequeues the next message, it is the `COMMIT` or `ROLLBACK` that thread 1 is waiting on.

Thread 2 finishes that command and moves onto the next, which is a `BEGIN`, and it blocks waiting for a `COMMIT` or `ROLLBACK`.

Thread 1 finishes, because it was unblocked by thread 2's first command, and gets the next command. It is the `COMMIT` or `ROLLBACK` that thread 2 is now waiting on.